klippy/extras/bltouch.py

```diff
@@ -53,54 +59,96 @@ def __init__(self, config):
-        # Raise the bltouch probe and test if probe is raised
-        self.send_cmd(print_time, 'reset')
-        home_time = print_time + self.pin_move_time
-        self.send_cmd(home_time, 'touch_mode')
-        self.send_cmd(home_time + MIN_CMD_TIME, None)
-        # Perform endstop check to verify bltouch reports probe raised
+        if self.next_cmd_time > print_time:
+            toolhead.dwell(self.next_cmd_time - print_time)
+        else:
+            self.next_cmd_time = print_time
+    def send_cmd(self, cmd, duration=MIN_CMD_TIME):
+        self.mcu_pwm.set_pwm(self.next_cmd_time, Commands[cmd] / SIGNAL_PERIOD)
+        # Translate duration to ticks to avoid any secondary mcu clock skew
+        mcu = self.mcu_pwm.get_mcu()
+        cmd_clock = mcu.print_time_to_clock(self.next_cmd_time)
+        cmd_clock += mcu.seconds_to_clock(max(duration, MIN_CMD_TIME))
+        self.next_cmd_time = mcu.clock_to_print_time(cmd_clock)
+        return self.next_cmd_time
+    def verify_state(self, check_start_time, check_end_time, triggered, msg):
+        # Perform endstop check to verify bltouch reports desired state
         prev_positions = [s.get_commanded_position()
                           for s in self.mcu_endstop.get_steppers()]
-        self.mcu_endstop.home_start(home_time, ENDSTOP_SAMPLE_TIME,
-                                    ENDSTOP_SAMPLE_COUNT, ENDSTOP_REST_TIME)
+        self.mcu_endstop.home_start(check_start_time, ENDSTOP_SAMPLE_TIME,
+                                    ENDSTOP_SAMPLE_COUNT, ENDSTOP_REST_TIME,
+                                    triggered=triggered)
         try:
-            self.mcu_endstop.home_wait(home_time + MIN_CMD_TIME)
+            self.mcu_endstop.home_wait(check_end_time)
         except self.mcu_endstop.TimeoutError as e:
-            raise homing.EndstopError("BLTouch sensor test failed")
+            raise homing.EndstopError("BLTouch failed to %s" % (msg,))
         for s, pos in zip(self.mcu_endstop.get_steppers(), prev_positions):
             s.set_commanded_position(pos)
+    def raise_probe(self):
+        self.send_cmd('reset')
+        check_start_time = self.send_cmd('pin_up', duration=self.pin_move_time)
+        check_end_time = self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time,
+                          self.pin_up_triggered, "raise probe")
+    def test_sensor(self):
+        if not self.pin_up_touch_triggered or self.pin_up_triggered:
+            # Nothing to test
+            return
+        toolhead = self.printer.lookup_object('toolhead')
+        print_time = toolhead.get_last_move_time()
+        if print_time < self.next_test_time:
+            self.next_test_time = print_time + TEST_TIME
+            return
+        # Raise the bltouch probe and test if probe is raised
+        self.sync_print_time()
+        check_start_time = self.send_cmd('reset', duration=self.pin_move_time)
+        check_end_time = self.send_cmd('touch_mode')
```

Comment:
The following line (not quoted) turns off the pwm.  A successful sensor test is always immediately followed by a pin_down command.

Reply:
Touch mode stays on until you reset the probe. It doesn't operate in a mode that requires a constant PWM signal for the modes and commands, as already demonstrated by the need to set it to zero after the commands. It's more like a simple command protocol, which happens to use the pulse lengths for the commands.

Reply:
That only applies if you assume that a `pin_down` will end the touch mode, which I wouldn't bet on. In fact my initial assumption was that only a `reset` can do that, since the documentation calls it a mode, which in my ears sounds like something that stays on until you disable it.  I did however test it quickly and at least on my genuine BLTouch Smart probe, it indeed puts the probe into normal mode again. 

Still, I don't see any harm in adding an extra `reset` command there, because we know that clones handle the touch mode differently and the additional reset should make sure it works on a wider range of probes, both on probes that need a reset and probes that automatically disables the mode on the next command.